Keep track of the derivative for unit_time

basnijholt opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
The derivative component is still in beta.

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

With this implementation, we will get a better estimate of the derivate during the timescale that is relevant to the sensor (indicated by the new option `time_window`).

This solves a problem where sensors have a low output resolution.
For example, a temperature sensor that can only report integer numbers (or half-integer).
It might report values every few seconds that are the same (e.g. 18 deg) and then suddenly go up one value (to 19 deg).
**Only** at that moment (with the current implementation) the derivative will be finite, otherwise, it will be zero. This instantaneous derivative isn't a good estimate of the true derivative.

The problem is captured by the following plot:
```python
import matplotlib.pyplot as plt
import numpy as np
xs = np.linspace(0, 2 * np.pi)
true_sensor = 4 * np.sin(xs)
sensor_output = np.round(true_sensor, 0)
plt.plot(xs, true_sensor, label='true sensor')
plt.plot(xs, sensor_output, label='sensor value')
plt.scatter(xs, sensor_output, c="C1")
plt.plot(xs, np.gradient(sensor_output), label='calculated derivative (with current method)')
plt.plot(xs, 5*np.cos(xs), label='true derivative')
plt.legend()
```

![image](https://user-images.githubusercontent.com/6897215/73616114-8e0ea680-460f-11ea-92c2-2391c673bf48.png)

With my proposed implementation, this problem will not occur, because it takes the average derivative of `time_window`.

I have added a `time_window` option, if you leave it out or if you set it to 0, the behavior is as it is in the `dev` branch.

Config example:
```
sensor:
  - platform: derivative
    source: sensor.gas_consumption
    name: Gas per hour
    unit: m³/h
    unit_time: h
    time_window: "00:30:00"
```
will average over the last half hour.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/11963

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for https://github.com/home-assistant/home-assistant.io/pull/11963
If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

basnijholt committed:
keep track of the derivative for unit_time

In this way, you will get a better estimate of the derivate during
the timescale that is relavant to the sensor.

This solved a problem where sensors have a low output resolution.
For example a temperature sensor that can only be integer numbers.

It might report many values that are the same and then suddenly go up one value.
Only in that moment (with the current implementation) the derivative will be finite.

With my proposed implementation, this problem will not occur, because it takes the average
derivative of the last `unit_time`.

basnijholt committed:
only loop as much as needed

basnijholt committed:
treat the special case of 1 entry

basnijholt commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -114,6 +115,25 @@ def calc_derivative(entity, old_state, new_state):
+
+            # Get indices of tuples that are older than `unit_time`
+            to_remove = []
+            for i, (timestamp, _) in enumerate(self._state_list):
```
I can replace this logic with
```
            self._state_list = [
                (timestamp, state)
                for (timestamp, state) in self._state_list
                if (now - timestamp).total_seconds() < self._unit_time
            ]
```
but it would be less efficient, but that might not matter?

afaucogney commented:
> With this implementation, we will get a better estimate of the derivate during the timescale that is relevant to the sensor (indicated by unit_time).

unit_time is not about relevance, this is a part of the unit of the derivative result, that you can calculate for different time scale.

basnijholt commented:
Then providing a `average_over` argument makes sense. If the user doesn’t set or, or sets it to `None`, the original behaviour is preserved.

> On 2 Feb 2020, at 19:00, Faucogney Anthony <notifications@github.com> wrote:
> 
> With this implementation, we will get a better estimate of the derivate during the timescale that is relevant to the sensor (indicated by unit_time).
> 
> unit_time is not about relevance, this is a part of the unit of the derivative result, that you can calculate for different time scale.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/home-assistant/home-assistant/pull/31397?email_source=notifications&email_token=ABUT4P2YQMOKENNRJQHNGE3RA4C5RA5CNFSM4KOWRE42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKR44TI#issuecomment-581160525>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABUT4P6UYPWFEXBD744N4FLRA4C5RANCNFSM4KOWRE4Q>.
>

basnijholt commented:
Then providing a `average_over` argument makes sense. If the user doesn’t set  it (sets it to `None`) or ,`average_over=0` the original behavior is preserved.

basnijholt committed:
add option time_window

basnijholt commented:
I have added a `time_window` option, if you leave it out or if you set it to 0, the behaviour is as it is in the `dev` branch.

It is in the same unit as `time_unit`, so a config like:
```
sensor:
  - platform: derivative
    source: sensor.gas_consumption
    name: Gas per hour
    unit: m³/h
    unit_time: h
    time_window: 0.5
```
will average over the last half hour.

springstan commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -93,6 +98,10 @@ def __init__(
         self._unit_prefix = UNIT_PREFIXES[unit_prefix]
         self._unit_time = UNIT_TIME[unit_time]
+        if time_window is None:
+            self._time_window = 0
```
Instead of specifying a default value here, we should declare a `DEFAULT_TIME_WINDOW` in the config schema and set it to `0`.

basnijholt replied:
Done!

springstan commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -114,6 +123,25 @@ def calc_derivative(entity, old_state, new_state):
+            now = new_state.last_updated
+            self._state_list.append((now, new_state.state))
+
+            # Get indices of tuples that are older than `unit_time`
```
Aren't we using the `time_window` in this loop?
```suggestion
            # Get indices of tuples that are older and outside of the `time_window`
```

basnijholt replied:
Fixed! Thanks for the suggestion 👍

balloob commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -46,6 +47,7 @@
         vol.Optional(CONF_UNIT_PREFIX, default=None): vol.In(UNIT_PREFIXES),
         vol.Optional(CONF_UNIT_TIME, default="h"): vol.In(UNIT_TIME),
         vol.Optional(CONF_UNIT): cv.string,
+        vol.Optional(CONF_TIME_WINDOW): vol.Coerce(float),
```
I think that we should use `cv.time_period` ?

basnijholt replied:
Renamed :)

balloob commented:
Don't forget to add a test.

basnijholt committed:
use cv.time_period

basnijholt committed:
fix comment

basnijholt committed:
set time_window=0 by default

basnijholt committed:
rephrase comment

basnijholt committed:
use timedelta for time_window

basnijholt committed:
fix the "G" unit_prefix and add more prefixes

https://en.wikipedia.org/wiki/Unit_prefix

basnijholt committed:
simplify logic

basnijholt committed:
add debugging lines

basnijholt committed:
fix bug where the there was a division of unit_time instead of multiplication

basnijholt committed:
simplify tests

basnijholt committed:
add test_data_moving_average_for_discrete_sensor

basnijholt committed:
fix test_dataSet6

basnijholt commented:
@balloob, I have added a test! 👍 

I've also refactored the tests and found an actual bug and fixed it (see this commit https://github.com/home-assistant/home-assistant/pull/31397/commits/243d75c991c118df22be9e16b5bbe1979fa2b905).

There was a test that confirmed the broken state, which I've fixed in 3f9fcac.

basnijholt committed:
improve readability of the tests

basnijholt committed:
better explain the test

basnijholt committed:
remove debugging log lines

springstan reviewed (approved):
Looks good to me, especially the tests look pretty clean now 👌🎉

afaucogney commented:
@basnijholt, I'm happy you have reviewed deeply the derivative integration. I'm neither a "python guy", neither a "derivative guy".

However I still have some open question:
1/ First I update you plot with same proportional factor for both derivatives (to get the same plot)
![Capture d’écran 2020-02-04 à 10 24 20](https://user-images.githubusercontent.com/6554201/73731549-fa38f980-4738-11ea-81e5-ad106699f4b2.png)

=> it looks better to me, and will provide less confusion. (with the first chart, it seems the original is almost always zero +- something)

- First, it would be good to show you result, (that we can compare with original version)
- Then, there is something I do not understand. You may see that the gradient of the sensor_value is high (always higher of the true_sensor if > 0) or 0.

**=> From this fact, how the true_derivative can be higher (in value) than the calculated one (the original one) ?**

_PS: As I said at the beginning of the "derivative" PR, this code is based on the "integration" integration. So maybe you should have a look on it, as you discovered many findings here !_

afaucogney commented:
I played a bit with jupyter, then this should more realistic to me.

![Capture d’écran 2020-02-04 à 11 02 31](https://user-images.githubusercontent.com/6554201/73734449-f2c81f00-473d-11ea-94b3-b40aebe4b437.png)

basnijholt commented:
Hi @afaucogney, you are right! I made a mistake in my plot. I both used `5*np.cos(xs)` instead of `4*np.cos(xs)`, so the derivative shouldn't be larger than the function itself.

Also, as you have correctly done, in my plot I didn't divide by Δx.

Your analysis shows, that if you have a densely sampled discrete sensor, that the derivate gives high peaks (which will become delta peaks in the limit of `Δx → 0`). That is the exact problem that I hoped to solve :)

I am doing some more small checks now.

Since now we essentially do a moving average, the derivative that we calculate is actually of `t_now - time_window/2`. I think we can still easily correct that too though.
